### ISSUE TITLE
Up/Down arrow keys  get "stuck" after pressing another arrow key at the same time in VoiceOver (Safari)

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/balloons-and-static-electricity#384]
Steps to reproduce:
 - Grab balloon with keyboard.
- Press and hold "up" arrow key
- Press and hold "right" arrow key
- Release both keys

Balloon will continue to move up. Only happens for vertical movement with arrow keys.

**Comment by jessegreenberg** [comment]
It is as though the `keyup` event for the vertical  arrow keys don't reach the if a horizontal arrow key is pressed at the same time. Problem does not occur when pressing vertical arrow key and then a letter key.

**Comment by jessegreenberg** [comment]
What is happening is even weirder. I see two keyup events for the horizontal arrow keys. Watching the DOM events, here is the rundown:

- Press and hold "up" arrow key <---DOM receives a keydown event for the up arrow key
- Press and hold "right" arrow key <-- DOM receives a keydown event for the right arrow key
- Release the "up arrow key <--- DOM receives a keyup event for the **right** arrow key
- Release the "right" arrow key <---DOM receives a keyup event for the **right arrow key

**Comment by jessegreenberg** [comment]
I am going to try to reproduce in a context outside of the sim.

**Comment by jessegreenberg** [comment]
Note that the order of horizontal and vertical keys matter. This problem only surfaces if you press the vertical arrow keys before the horizontal arrow keys.

**Comment by jessegreenberg** [comment]
I am seeing the exact same behavior in this JSFiddle. https://jsfiddle.net/r39rg5zL/8/. With VoiceOver turned on following steps in https://github.com/phetsims/balloons-and-static-electricity/issues/384#issuecomment-378357582 I print the following events

keydown: 38 (UpArrowKey)
keydown: 39 (RightArrowKey)
keyup: 39 (RightArrowKey)
keyup: 39 (RightArrowKey)

And so the sim thinks the up arrow key is still held down, and the balloon continues to move up. This will be a problem for all sims that use arrow keys to drag in 2D.

**Comment by jessegreenberg** [comment]
I can't think of a good fix for this. Perhaps the best thing to do is gracefully handle it. How about if we receive a keyup event for a key that is not currently down, we make the KeyboardDragListener think that  no arrow keys are down.

**Comment by jessegreenberg** [comment]
That doesn't work if you continue to hold down a horizontal arrow key after releasing the vertical arrow key.

**Comment by jessegreenberg** [comment]
I  don't think  this is possible to detect when pressing and holding keys for longer than the browser's press and hold delay because by then there are  matching horizontal keydown events for the errant keyup event.

**Comment by jessegreenberg** [comment]
I tried just only allowing one arrow key to be pressed  at a time, but doesn't get to the root of the problem. The issue is that even if we prevent adding behavior to the second arrow key, VoiceOver tells the browser that the *second* arrow key was released rather than the first one and we are left with the same problem.

**Comment by jessegreenberg** [comment]
Here is one workaround that seems to keep the sim working:

If we are in Safari, and more than one arrow key is pressed down at a time, we stop all keyboard dragging and reset `KeyboardDragListener` so that it thinks that no keys are pressed down. This will fix this VoiceOver bug gracefully,  and dragging can continue next time an arrow key is pressed. Unfortunately, this will make it impossible to move the object diagonally with the arrow keys in Safari. Movement with the WASD keys will work as they always have.

We will also submit a bug report to Apple.

@emily-phet @terracoda I am a bit over an hour into searching for a workaround for this bug. Does this proposal seem appropriate to you?

**Comment by emily-phet** [comment]
@jessegreenberg if you press any arrow key once the errant up behavior is happening, does this stop the motion?

**Comment by jessegreenberg** [comment]
With the workaround, that is correct - motion stops.

**Comment by jessegreenberg** [comment]
Here is a version with this fix if it helps to see it in action: https://www.colorado.edu/physics/phet/dev/html/balloons-and-static-electricity/1.4.0-dev.20/phet/balloons-and-static-electricity_en_phet.html

I could also share screens and voice chat to talk in more detail if that would be helpful.

**Comment by emily-phet** [comment]
@jessegreenberg So without the workaround, the motion continues until...when? How do you stop it (without a workaround)?

**Comment by jessegreenberg** [comment]
Without the workaround, the balloon will keep moving until it hits the top or bottom of the Play Area (constrained by the model). But the sim thinks that the arrow key is still pressed down after it has been released, so if you try to move the balloon down again, it will just move back up to the top of the Play Area once you have released the "down" arrow key.

**Comment by emily-phet** [comment]
@jessegreenberg Okay, that's helpful. I was wondering if it might be a situation where when the problem arises, some of the natural troubleshooting a person might do may stop the problem. Sounds like that's not the case here. You're workaround seems reasonable to me.

**Comment by terracoda** [comment]
@jessegreenberg , I agree with @emily-phet, this solution seems reasonable. It only affects Safari, and is it safe to assume that diagonal dragging is still possible with WASD keys?

**Comment by jessegreenberg** [comment]
> and is it safe to assume that diagonal dragging is still possible with WASD keys?

Correct! Thanks @terracoda and  @emily-phet,  in that case I am closing this issue.

**Comment by lmulhall-phet** [comment]
> and is it safe to assume that diagonal dragging is still possible with WASD keys?

This is not the case for https://bayes.colorado.edu/dev/html/friction/1.5.0-dev.6/phet/friction_all_phet.html as noted in phetsims/friction/issues/79.

@jessegreenberg, do we care? I know this isn't a huge issue...

**Comment by lmulhall-phet** [comment]
Ah, nevermind! I was thinking arrow keys. Sorry, folks!
